src/coreclr/jit/lower.cpp

```diff
@@ -3654,12 +3654,27 @@ void Lowering::LowerRetStruct(GenTreeUnOp* ret)
             break;
 
         case GT_OBJ:
-            retVal->ChangeOper(GT_IND);
-            FALLTHROUGH;
         case GT_IND:
+        {
+            // Spill to a local if sizes don't match so we can avoid the "load more than requested"
+            // problem, e.g. struct size is 5 and we emit "ldr x0, [x1]"
+            unsigned realSize = retVal->OperIs(GT_OBJ) ? retVal->AsObj()->Size() : retVal->AsIndir()->Size();
+            if (genTypeSize(nativeReturnType) != realSize && realSize != 0)
```

Comment:
It is still a problem if `realSize` is zero - it means there was a `GT_FIELD` node under the return, which could have also been for a small struct. To know the real size of such an `IND<struct>` you'd need to query `info.compMethodInfo->args.retTypeClass`). If you're going to go with this (instead of accepting the regression for small struct fields), it would be good to add `TODO-ADDR: delete once "IND<struct>" nodes are no more` there.

Reply:
Let's go with the 2nd option then

Reply:
To be clear, I meant that the `&& realSize != 0` check makes it a correctness issue as-is. Something like below:
```c
SmallStruct Get(LargeStruct* p)
{
    return p->SmallStructField;
}
```
Will load out-of-bounds.